src/ripple/unl/tests/qalloc.h

```diff
@@ -0,0 +1,396 @@
+
+    void
+    destroy (T* t);
+
+    template <class U>
+    bool
+    operator== (qalloc_type<U> const& u);
+
+    template <class U>
+    bool
+    operator!= (qalloc_type<U> const& u);
+};
+
+/** Allocator optimized for delete in temporal order.
+
+    This allocator is optimized for the case where objects
+    are deleted in approximately the same order that they
+    were created.
+*/
+using qalloc = qalloc_type<int>;
+
+//------------------------------------------------------------------------------
+
+namespace detail {
+
+template <class _>
+qalloc_impl<_>::block::block (std::size_t bytes)
+    : bytes_ (bytes - sizeof(*this))
+    , remain_ (bytes_)
+    , free_ (reinterpret_cast<
+        std::uint8_t*>(this + 1))
+{
+}
+
+template <class _>
+void*
+qalloc_impl<_>::block::allocate(
+    std::size_t bytes, std::size_t align)
+{
+//align = 16;
+    align = std::max(align,
+        std::alignment_of<block*>::value);
+    auto pad = [](void const* p, std::size_t align)
+    {
+        auto const i = reinterpret_cast<
+            std::uintptr_t>(p);
+        return (align - (i % align)) % align;
+    };
+
+    auto const n0 =
+        pad(free_ + sizeof(block*), align);
+    auto const n1 =
+        n0 + sizeof(block*) + bytes;
+    if (remain_ < n1)
+        return nullptr;
+    auto p = reinterpret_cast<block**>(
+        free_ + n0 + sizeof(block*));
+    assert(pad(p - 1,
+        std::alignment_of<block*>::value) == 0);
+    p[-1] = this;
+    ++count_;
+    free_ += n1;
+    remain_ -= n1;
+    return p;
+}
+
+template <class _>
+bool
+qalloc_impl<_>::block::deallocate()
+{
+    --count_;
+    if (count_ > 0)
+        return false;
+    remain_ = bytes_;
+    free_ = reinterpret_cast<
+        std::uint8_t*>(this + 1);
+    return true;
+}
+
+template <class _>
+template <class U>
+inline
+auto
+qalloc_impl<_>::block::stash (U const* u) ->
+    block*&
+{
+    static auto const a =
+        std::alignment_of<block*>::value;
+    return *reinterpret_cast<block**>(
+        a * ((reinterpret_cast<std::uintptr_t>(u) - sizeof(block*)) / a));
+}
+
+template <class _>
+qalloc_impl<_>::~qalloc_impl()
+{
+    if (used_)
+        std::free(used_);
+    while (free_)
+    {
+        auto const next = free_->next;
+        free_->~block();
+        std::free(free_);
+        free_ = next;
+    }
+}
+
+template <class _>
+void*
+qalloc_impl<_>::allocate(
+    std::size_t bytes, std::size_t align)
+{
+    if (used_)
+    {
+        auto const p =
+            used_->allocate(bytes, align);
+        if (p)
+            return p;
+        used_ = nullptr;
+    }
+    if (free_)
+    {
+        auto const p =
+            free_->allocate(bytes, align);
+        if (p)
+        {
+            used_ = free_;
+            free_ = free_->next;
+            return p;
+        }
+    }
+    // VFALCO Formula needs audit
+    auto const n = std::max<std::size_t>(
+        block_size, sizeof(block) +
+            sizeof(block*) + bytes + 2 * align);
```

Comment:
I don't see any padding between `block` and `block*` due to the nature of `block` being composed of "word types".  However if we have such a concern, perhaps we could move `block*` into `block`?